src/main/java/org/opensearch/neuralsearch/processor/combination/ScoreCombiner.java

```diff
@@ -318,4 +318,75 @@ private TotalHits getTotalHits(final List<TopDocs> topDocsPerSubQuery, final lon
         }
         return new TotalHits(maxHits, totalHits);
     }
+
+    /**
+     * Explain the score combination technique for each document in the given queryTopDocs.
+     * @param queryTopDocs
+     * @param combinationTechnique
+     * @param sort
+     * @return a map of SearchShard and List of ExplainationDetails for each document
+     */
+    public Map<SearchShard, List<ExplanationDetails>> explain(
+        final List<CompoundTopDocs> queryTopDocs,
+        final ScoreCombinationTechnique combinationTechnique,
+        final Sort sort
+    ) {
+        // In case of duplicate keys, keep the first value
+        Map<SearchShard, List<ExplanationDetails>> explanations = new HashMap<>();
+        for (CompoundTopDocs compoundQueryTopDocs : queryTopDocs) {
+            explanations.putIfAbsent(
+                compoundQueryTopDocs.getSearchShard(),
+                explainByShard(combinationTechnique, compoundQueryTopDocs, sort)
+            );
+        }
+        return explanations;
+    }
+
+    private List<ExplanationDetails> explainByShard(
+        final ScoreCombinationTechnique scoreCombinationTechnique,
+        final CompoundTopDocs compoundQueryTopDocs,
+        final Sort sort
+    ) {
+        if (Objects.isNull(compoundQueryTopDocs) || compoundQueryTopDocs.getTotalHits().value == 0) {
+            return List.of();
+        }
+        // create map of normalized scores results returned from the single shard
+        Map<Integer, float[]> normalizedScoresPerDoc = getNormalizedScoresPerDocument(compoundQueryTopDocs.getTopDocs());
+        // combine scores
+        Map<Integer, Float> combinedNormalizedScoresByDocId = normalizedScoresPerDoc.entrySet()
+            .stream()
+            .collect(Collectors.toMap(Map.Entry::getKey, entry -> scoreCombinationTechnique.combine(entry.getValue())));
+        // sort combined scores as per sorting criteria - either score desc or field sorting
+        Collection<Integer> sortedDocsIds = getSortedDocsIds(compoundQueryTopDocs, sort, combinedNormalizedScoresByDocId);
+
+        List<ExplanationDetails> listOfExplanations = sortedDocsIds.stream()
```

Comment:
Please avoid stream here. Can you use standard for loop

Reply:
ack